File: Python/symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param {TreeNode} root
    # @return {boolean}
    def isSymmetric(self, root):
        def isSymmetric(n1, n2):
            if not n1: return n2==None
            if not n2: return False
            if n1.val!=n2.val: return False
            if not isSymmetric(n1.left, n2.right): return False
            if not isSymmetric(n1.right, n2.left): return False
            return True
        return isSymmetric(root, root)

File: C++/symmetric-tree.cpp
/**
 *  * Definition for a binary tree node.
 *   * struct TreeNode {
 *    *     int val;
 *     *     TreeNode *left;
 *      *     TreeNode *right;
 *       *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 *        * };
 *         */
class Solution {
private:
    bool isSymmetric(TreeNode* n1, TreeNode* n2) {
    if(!n1) return !n2;
    if(!n2) return false;
    if(n1->val!=n2->val) return false;
    if(!isSymmetric(n1->left, n2->right)) return false;
    if(!isSymmetric(n1->right, n2->left)) return false;
    return true;
}
public:
    bool isSymmetric(TreeNode* root) {
        return isSymmetric(root, root);
    }
};

